ui/lib/src/nvui/handler.ts

```diff
@@ -81,62 +81,124 @@ export function arrowKeyHandler(pov: Color, borderSound: () => void) {
     if ($moveBox.val() === '') {
-      // if user selects another's piece first
-      if ($evBtn.attr('color') === opponentColor) return;
-      // as long as the user is selecting a piece and not a blank tile
-      if ($evBtn.text().match(/^[^\-+]+/g)) {
+      if (
+        $evBtn.attr('color') === opponentColor ||
+        $evBtn.attr('piece') === '-' ||
+        $evBtn.attr('piece') === '+'
+      ) {
+        $boardLive.text(keyText(ev.target as HTMLElement) + ' not selectable');
+      } else {
         $moveBox.val(pos);
-        selectSound();
+        clear('selection');
+        $evBtn.addClass('selected');
+        $evBtn.text($evBtn.attr('text') + ' selected');
       }
     } else {
-      const first = $moveBox.val();
-      if (typeof first !== 'string' || !isKey(first)) return;
-      const $firstPiece = $(squareSelector(first[1], first[0]));
-      $moveBox.val($moveBox.val() + pos);
-      // this is coupled to pieceJumpingHandler() noticing that the attribute is set and acting differently. TODO: make cleaner
-      // if pawn promotion
-      if (rank === promotionRank && $firstPiece.attr('piece')?.toLowerCase() === 'p') {
-        $evBtn.attr('promotion', 'true');
-        $boardLive.text('Promote to? q for queen, n for knight, r for rook, b for bishop');
-        return;
+      const input = $moveBox.val();
+      if (typeof input !== 'string') return;
+      if (isKey(input)) {
+        const $firstPiece = $(squareSelector(input[1], input[0]));
+        $moveBox.val($moveBox.val() + pos);
+        // this is coupled to pieceJumpingHandler() noticing that the attribute is set and acting differently.
+        if (rank === promotionRank && file && $firstPiece.attr('piece')?.toLowerCase() === 'p') {
+          $evBtn.attr('promotion', 'true');
+          if (!isTouchDevice)
+            $boardLive.text('Promote to: q for queen, n for knight, r for rook, b for bishop');
+          else {
+            const queenPromotionKey = $(squareSelector(promotionRank === '8' ? '8' : '1', file));
+            const knightPromotionKey = $(squareSelector(promotionRank === '8' ? '7' : '2', file));
+            const rookPromotionKey = $(squareSelector(promotionRank === '8' ? '6' : '3', file));
+            const bishopPromotionKey = $(squareSelector(promotionRank === '8' ? '5' : '4', file));
+            const cancelPromotionKey = $(squareSelector(promotionRank === '8' ? '4' : '5', file));
+            queenPromotionKey.attr('promoteTo', 'q');
+            queenPromotionKey.text('promote to queen');
+            knightPromotionKey.attr('promoteTo', 'n');
+            knightPromotionKey.text('promote to knight');
+            rookPromotionKey.attr('promoteTo', 'r');
+            rookPromotionKey.text('promote to rook');
+            bishopPromotionKey.attr('promoteTo', 'b');
+            bishopPromotionKey.text('promote to bishop');
+            cancelPromotionKey.attr('promoteTo', 'x');
+            cancelPromotionKey.text('cancel');
```

Comment:
This can and should be easily golfed down. This also ignores antichess promotion to king. So you'll need to account for variant in this function. Otherwise very nice!